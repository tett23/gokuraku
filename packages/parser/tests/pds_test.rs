extern crate parser;

use insta::{assert_debug_snapshot, glob};
use std::fs;

#[test]
fn pds_test() {
    glob!("../fixtures/pds", "*.pds", |path| {
        let input = fs::read_to_string(path).unwrap();
        let ast = parser::prose_down_script_parse(&input).unwrap();

        assert_debug_snapshot!(ast);
    });
}
