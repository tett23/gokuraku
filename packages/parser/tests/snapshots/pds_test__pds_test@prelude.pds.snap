---
source: packages/parser/tests/pds_test.rs
expression: ast
input_file: packages/parser/fixtures/pds/prelude.pds
---
Module {
    statements: [
        InstDef(
            InstDef {
                ident: InstIdent(
                    "@atomic",
                ),
                expr: TypeExpr {
                    constraints: [],
                    expr: Literal(
                        Tuple(
                            0,
                            [],
                        ),
                    ),
                },
            },
        ),
        InstDef(
            InstDef {
                ident: InstIdent(
                    "@write",
                ),
                expr: TypeExpr {
                    constraints: [],
                    expr: Arrow(
                        Literal(
                            Ident(
                                TypeIdent(
                                    "Char",
                                ),
                            ),
                        ),
                        Literal(
                            Context(
                                TypeIdent(
                                    "IO",
                                ),
                                Literal(
                                    Tuple(
                                        0,
                                        [],
                                    ),
                                ),
                            ),
                        ),
                    ),
                },
            },
        ),
        InstDef(
            InstDef {
                ident: InstIdent(
                    "@read",
                ),
                expr: TypeExpr {
                    constraints: [],
                    expr: Arrow(
                        Literal(
                            Tuple(
                                0,
                                [],
                            ),
                        ),
                        Literal(
                            Context(
                                TypeIdent(
                                    "IO",
                                ),
                                Literal(
                                    Ident(
                                        TypeIdent(
                                            "Char",
                                        ),
                                    ),
                                ),
                            ),
                        ),
                    ),
                },
            },
        ),
        DataAssign(
            DataAssign {
                modifiers: [
                    Nominal,
                ],
                constraints: [],
                ident: TypeIdent(
                    "True",
                ),
                args: [],
                expr: Value(
                    Unit,
                ),
            },
        ),
        DataAssign(
            DataAssign {
                modifiers: [
                    Nominal,
                ],
                constraints: [],
                ident: TypeIdent(
                    "False",
                ),
                args: [],
                expr: Value(
                    Unit,
                ),
            },
        ),
        DataAssign(
            DataAssign {
                modifiers: [],
                constraints: [],
                ident: TypeIdent(
                    "Bool",
                ),
                args: [],
                expr: Or(
                    Value(
                        TypeIdent(
                            TypeIdent(
                                "True",
                            ),
                        ),
                    ),
                    Value(
                        TypeIdent(
                            TypeIdent(
                                "False",
                            ),
                        ),
                    ),
                ),
            },
        ),
        DataAssign(
            DataAssign {
                modifiers: [],
                constraints: [],
                ident: TypeIdent(
                    "Option",
                ),
                args: [
                    TypeIdent(
                        "a",
                    ),
                ],
                expr: Or(
                    Value(
                        Context(
                            TypeIdent(
                                "Some",
                            ),
                            [
                                TypeIdent(
                                    "a",
                                ),
                            ],
                        ),
                    ),
                    Value(
                        TypeIdent(
                            TypeIdent(
                                "Nil",
                            ),
                        ),
                    ),
                ),
            },
        ),
        DataAssign(
            DataAssign {
                modifiers: [],
                constraints: [],
                ident: TypeIdent(
                    "Either",
                ),
                args: [
                    TypeIdent(
                        "a",
                    ),
                    TypeIdent(
                        "b",
                    ),
                ],
                expr: Or(
                    Value(
                        Context(
                            TypeIdent(
                                "Left",
                            ),
                            [
                                TypeIdent(
                                    "a",
                                ),
                            ],
                        ),
                    ),
                    Value(
                        Context(
                            TypeIdent(
                                "Right",
                            ),
                            [
                                TypeIdent(
                                    "b",
                                ),
                            ],
                        ),
                    ),
                ),
            },
        ),
        DataAssign(
            DataAssign {
                modifiers: [],
                constraints: [],
                ident: TypeIdent(
                    "Result",
                ),
                args: [
                    TypeIdent(
                        "a",
                    ),
                    TypeIdent(
                        "b",
                    ),
                ],
                expr: Or(
                    Value(
                        Context(
                            TypeIdent(
                                "Ok",
                            ),
                            [
                                TypeIdent(
                                    "a",
                                ),
                            ],
                        ),
                    ),
                    Value(
                        Context(
                            TypeIdent(
                                "Err",
                            ),
                            [
                                TypeIdent(
                                    "b",
                                ),
                            ],
                        ),
                    ),
                ),
            },
        ),
        TraitDef(
            TraitDef {
                constraints: [],
                ident: TypeIdent(
                    "Show",
                ),
                args: [
                    TypeIdent(
                        "a",
                    ),
                ],
                where_clause: Module {
                    statements: [
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "show",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "Text",
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                    ],
                },
            },
        ),
        DataAssign(
            DataAssign {
                modifiers: [],
                constraints: [],
                ident: TypeIdent(
                    "Ordering",
                ),
                args: [],
                expr: Or(
                    Value(
                        TypeIdent(
                            TypeIdent(
                                "LT",
                            ),
                        ),
                    ),
                    Or(
                        Value(
                            TypeIdent(
                                TypeIdent(
                                    "EQ",
                                ),
                            ),
                        ),
                        Value(
                            TypeIdent(
                                TypeIdent(
                                    "GT",
                                ),
                            ),
                        ),
                    ),
                ),
            },
        ),
        TraitDef(
            TraitDef {
                constraints: [],
                ident: TypeIdent(
                    "PartialOrd",
                ),
                args: [
                    TypeIdent(
                        "a",
                    ),
                    TypeIdent(
                        "b",
                    ),
                ],
                where_clause: Module {
                    statements: [
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "partialCmp",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Arrow(
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "b",
                                                    ),
                                                ),
                                            ),
                                            Literal(
                                                Context(
                                                    TypeIdent(
                                                        "Option",
                                                    ),
                                                    Literal(
                                                        Ident(
                                                            TypeIdent(
                                                                "Ordering",
                                                            ),
                                                        ),
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "(<)",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Arrow(
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "b",
                                                    ),
                                                ),
                                            ),
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "Bool",
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "(<=)",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Arrow(
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "b",
                                                    ),
                                                ),
                                            ),
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "Bool",
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "(>)",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Arrow(
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "b",
                                                    ),
                                                ),
                                            ),
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "Bool",
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "(>=)",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Arrow(
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "b",
                                                    ),
                                                ),
                                            ),
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "Bool",
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                    ],
                },
            },
        ),
        TraitDef(
            TraitDef {
                constraints: [],
                ident: TypeIdent(
                    "PartialEq",
                ),
                args: [
                    TypeIdent(
                        "a",
                    ),
                    TypeIdent(
                        "b",
                    ),
                ],
                where_clause: Module {
                    statements: [
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "(==)",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Arrow(
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "b",
                                                    ),
                                                ),
                                            ),
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "Bool",
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "(!=)",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Arrow(
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "b",
                                                    ),
                                                ),
                                            ),
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "Bool",
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                    ],
                },
            },
        ),
        TraitDef(
            TraitDef {
                constraints: [
                    Constraint {
                        ident: TypeIdent(
                            "PartialEq",
                        ),
                        args: [],
                    },
                ],
                ident: TypeIdent(
                    "Eq",
                ),
                args: [
                    TypeIdent(
                        "a",
                    ),
                ],
                where_clause: Module {
                    statements: [],
                },
            },
        ),
        TraitDef(
            TraitDef {
                constraints: [
                    Constraint {
                        ident: TypeIdent(
                            "PartialEq",
                        ),
                        args: [],
                    },
                    Constraint {
                        ident: TypeIdent(
                            "Eq",
                        ),
                        args: [],
                    },
                ],
                ident: TypeIdent(
                    "Ord",
                ),
                args: [
                    TypeIdent(
                        "a",
                    ),
                ],
                where_clause: Module {
                    statements: [
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "cmp",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Arrow(
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "a",
                                                    ),
                                                ),
                                            ),
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "Ordering",
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "max",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Arrow(
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "a",
                                                    ),
                                                ),
                                            ),
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "a",
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "min",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Arrow(
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "a",
                                                    ),
                                                ),
                                            ),
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "a",
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                        AssignDef(
                            AssignDef {
                                eta_envs: EtaEnvs(
                                    [],
                                ),
                                ident: Ident(
                                    "clamp",
                                ),
                                expr: TypeExpr {
                                    constraints: [],
                                    expr: Arrow(
                                        Literal(
                                            Ident(
                                                TypeIdent(
                                                    "a",
                                                ),
                                            ),
                                        ),
                                        Arrow(
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "a",
                                                    ),
                                                ),
                                            ),
                                            Literal(
                                                Ident(
                                                    TypeIdent(
                                                        "a",
                                                    ),
                                                ),
                                            ),
                                        ),
                                    ),
                                },
                            },
                        ),
                    ],
                },
            },
        ),
        ImplTrait(
            ImplTrait {
                constraints: [],
                ident: TypeIdent(
                    "Show",
                ),
                args: [
                    TypeIdent(
                        "Int",
                    ),
                ],
                where_clause: Module {
                    statements: [
                        LineComment(
                            LineComment(
                                "lineComment(768, 778)",
                            ),
                        ),
                        Assign(
                            Assign {
                                ident: Ident(
                                    "show",
                                ),
                                args: AssignArgs {
                                    patterns: [
                                        Bind(
                                            Ident(
                                                "a",
                                            ),
                                        ),
                                    ],
                                },
                                expr: Literal(
                                    Text(
                                        "",
                                    ),
                                ),
                                where_clause: Module {
                                    statements: [],
                                },
                            },
                        ),
                    ],
                },
            },
        ),
        ImplTrait(
            ImplTrait {
                constraints: [],
                ident: TypeIdent(
                    "Show",
                ),
                args: [
                    TypeIdent(
                        "Bool",
                    ),
                ],
                where_clause: Module {
                    statements: [
                        Assign(
                            Assign {
                                ident: Ident(
                                    "show",
                                ),
                                args: AssignArgs {
                                    patterns: [
                                        TypeIdent(
                                            TypeIdent(
                                                "True",
                                            ),
                                        ),
                                    ],
                                },
                                expr: Literal(
                                    Text(
                                        "True",
                                    ),
                                ),
                                where_clause: Module {
                                    statements: [],
                                },
                            },
                        ),
                        Assign(
                            Assign {
                                ident: Ident(
                                    "show",
                                ),
                                args: AssignArgs {
                                    patterns: [
                                        TypeIdent(
                                            TypeIdent(
                                                "False",
                                            ),
                                        ),
                                    ],
                                },
                                expr: Literal(
                                    Text(
                                        "False",
                                    ),
                                ),
                                where_clause: Module {
                                    statements: [],
                                },
                            },
                        ),
                    ],
                },
            },
        ),
    ],
}
